compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ReachabilityAnalyzer.java

```diff
@@ -628,7 +672,9 @@ public void visit(BLangDoClause doClause) {
                                                                                               env.enclInvokable));
 
         BLangBlockStmt body = doClause.body;
-        checkStatementExecutionValidity(body);
+        if (!(doClause.parent != null && doClause.parent.getKind() == NodeKind.DO_ACTION)) {
```

Comment:
Wouldn't it be clearer to say

```suggestion
        if (doClause.parent == null || doClause.parent.getKind() != NodeKind.DO_ACTION) {
```

Reply:
Why do we need this check in the first place?